Meteor AutoForm 7.1.0 - Autocomplete

4Z4T4R opened the pull request:
- Added new inputTypes autocomplete directory
- Added HTML for Autocomplete Template
- Added JS for Autocomplete Template
- Circular selection allows arrows to cycle tip to tail and vice versa
- Reactive options allow Tracker vars to provide labels and values to dropdown selections
- Simpl-Schema defined options provide the same path to values and labels as select lists.
- Navigable via arrows so your mouse can rest.
- No third party solutions so we're not feeling bloated.
- Uses bootstrap 4 dropdown class definitions for style so adopters can use theme packages, no problemo.
- Conforms to StandardJS delinting and format so that the gods (and developers) are happy.

4Z4T4R committed:
Introducing Autocomplete for Meteor AutoForm
- Added new inputTypes autocomplete directory
- Added HTML for Autocomplete Template
- Added JS for Autocomplete Template (WIP: Need actions!)
- Styles will go inside `meteor-autoform-themes` project

4Z4T4R committed:
Meteor AutoForm 7.1.0 - Native Autocomplete
- Circular selection
- Reactive options
- Simpl-Schema defined options
- Navigable via arrows
- No third party solutions

4Z4T4R committed:
Meteor AutoForm v7.1.0 - Autocomplete
- Uses bootstrap 4 dropdown class definitions for style
- Conforms to StandardJS delinting and format

jankapunkt commented on inputTypes/autocomplete/autocomplete.js:
```diff
@@ -0,0 +1,167 @@
+
+        // choose an answer on click
+        $suggestions.children().click((e) => {
+          const dataValue = $(e.target).attr('data-value')
```
can you please change this to

```javascript
const dataValue = me.$(e.target).attr('data-value')
``` 

so it uses the local-scoped jQuery selector plus it resolves the linter error.

4Z4T4R replied:
Done. Thx.

4Z4T4R committed:
Meteor AutoForm v7.1.0 - Autocomplete
- added hidden input field to allow labels and values in autocomplete; think 'Full Name' and 'UserId'
- now adjusting context for visible and hidden inputs, so AutoForm works as expected on submit
- run lint passed
- run test passed

4Z4T4R committed:
Meteor AutoForm v7.1.0 - Autocomplete
- Escape closes the dropdown
- Double-click opens the dropdown
- Removed a lone helpers definition of no value
- Preserved sort order of backing object array
- Form /should/ submit properly when 'Enter' triggers and dropdown is hidden
- Maybe we're all clear for launch, but I'm not counting my chickens yet

4Z4T4R committed:
Meteor AutoForm v7.1.0 - Autocomplete
- Fixed HTML structure to display bootstrap validation errors properly in autocomplete inputs.

4Z4T4R committed:
Meteor AutoForm v7.1.0 - Autocomplete
- Now the values between the hidden and visible fields remain identical so that validation works correctly if the value is deleted.

4Z4T4R committed:
Meteor AutoForm v7.1.0 - Autocomplete
- Now, after selecting from the autocomplete options, the field keeps focus so that one may tab to the next field.

jankapunkt committed:
inputTypes autocomplete update options reactively

jankapunkt committed:
inputTypes add missing ReactiveVar import

jankapunkt committed:
inputTypes autocomplete added explicit touch support

jankapunkt committed:
autocomplete ui and reactivity fixed

4Z4T4R commented on inputTypes/autocomplete/autocomplete.js:
```diff
@@ -230,4 +230,9 @@ Template.afAutocomplete.onRendered(function () {
     callback(e)
   })
+
+  // show on double click
```
change comment to `show on touch`
Good catch!

4Z4T4R commented on inputTypes/autocomplete/autocomplete.js:
```diff
@@ -121,6 +122,11 @@ Template.afAutocomplete.onRendered(function () {
     }
   })
 
+  const updateValue = value => {
+    $hidden.val(value)
+    $hidden.trigger('change')
+  }
+
```
Clever! Nicely done. Is it safe to assume that the change event triggers validation? on blur too?

4Z4T4R replied:
I see the blur change below.

jankapunkt replied:
The change is used to fire Updates for any reactive getter like AutoForm.getFieldValue

While blur triggers validation.

4Z4T4R commented on inputTypes/autocomplete/autocomplete.js:
```diff
@@ -214,13 +220,17 @@ Template.afAutocomplete.onRendered(function () {
     }
   }
 
+  $input.blur(() => {
+    $hidden.trigger('blur') // triggers re-validation
+  })
+
```
Ah! There it is. Well done. Thx!

4Z4T4R commented on inputTypes/autocomplete/autocomplete.js:
```diff
@@ -121,6 +122,11 @@ Template.afAutocomplete.onRendered(function () {
+
   // clear on blur?
   // TODO: Figure out how blur won't block "click"
   // $input.blur((e)=>{ clearDropdown(e) })
```
Remove this comment.

4Z4T4R reviewed (commented):
There's general feedback here. Just some curiosity around the trigger for the input/hidden input.

jankapunkt committed:
inputTypes autoComplete added comment on updateValue function

jankapunkt committed:
downgraded version to 7.0.0 until final release to avoid ci conflicts

jankapunkt committed:
inputTypes autocomplete corrected event bubbling on click and tab

jankapunkt committed:
inputTypes autocomplete emit change event only on empty (falsy) or valid (in selectOptons) value
